File: java/nbody/Force.java

package nbody;


public class Force {
    double x;
    double y;
    double z;

    Force(double _x) {
        x = _x;
        y = _x;
        z = _x;
    }

    Force(double _x, double _y, double _z) {
        x = _x;
        y = _y;
        z = _z;
    }

    public double norm2() {
        return x*x + y*y + z*z;
    }

    public Force dist(Force f2) {
        Force d = new Force(0.);
        d.x = x - f2.x;
        d.y = y - f2.y;
        d.z = z - f2.z;
        return d;
    }
}

File: julia/src/force.jl
module MyForce

export Force, norm2

type Force{T<:Real}
    x::T
    y::T
    z::T
end

Force{T<:Real}(x::T) = Force{T}(x, x, x)

convert{T<:Real}(::Type{Force{T}}, f::Force) = Force(convert(T, f.x), convert(T, f.y), convert(T, f.z))


norm2{T<:Real}(f::Force{T}) = f.x*f.x + f.y*f.y + f.z*f.z
norm{T<:Real}(f::Force{T}) = sqrt(f.x*f.x + f.y*f.y + f.z*f.z)

+{T<:Real}(a::Force{T}, b::Force{T}) = Force{T}(a.x+b.x, a.y+b.y, a.z+b.z)
-{T<:Real}(a::Force{T}, b::Force{T}) = Force{T}(a.x-b.x, a.y-b.y, a.z-b.z)
*{T<:Real}(a::Force{T}, b::Force{T}) = a.x*b.x, a.y*b.y, a.z*b.z

*{T<:Real}(a::Force{T}, b::T) = Force{T}(a.x*b, a.y*b, a.z*b)
*{T<:Real}(b::T, a::Force{T}) = a*b
/{T<:Real}(a::Force{T}, b::T) = Force{T}(a.x/b, a.y/b, a.z/b)


end

